### PR TITLE
Refactored bulk_get unit tests, implemented max_upload_size

### PR DESCRIPTION
**Pull request by chanchett**

### CODE REVIEW

File: test/bulk_get.cpp [hunk 1]
```diff
@@ -7,153 +7,263 @@
...
+      
+    ds3_free_available_chunks_response(chunk_response);
+    ds3_free_bulk_response(bulk_response);
+
+    clear_bucket(client, bucket_name);
+    free_client(client);
+}
+
+
+BOOST_AUTO_TEST_CASE( chunk_preference ) {
```
Comment (chanchett) [inline test/bulk_get.cpp]: This test doesn't check anything different from the normal bulk_get unit test, it's only checking that the call to set_preferred_number_of_chunks doesn't cause any errors.

File: src/ds3.c [hunk 2]
```diff
@@ -453,6 +453,20 @@ void ds3_request_set_max_keys(ds3_request* _request, uint32_t max_keys) {
     _set_query_param(_request, "max-keys", max_keys_s);
 }
 
+void ds3_request_set_preferred_number_of_chunks(ds3_request* _request, uint32_t num_chunks) {
+    char num_chunks_s[11];
```
Comment (DenverM80) [inline src/ds3.c]: What does this magic number represent?

I'm assuming its the length of 2^32 (4294967296) + NULL, which you should go ahead and declare:
static char* UNSIGNED_LONG_BASE_10 = "4294967296"
static unsigned char UNSIGNED_LONG_BASE_10_STR_LEN = 11;

File: test/bulk_get.cpp [hunk 3]
```diff
@@ -7,153 +7,263 @@
...
+    uint32_t result = 0;
+    for (uint32_t i = 0; i < num_files; i++) {
+        result += results[i].num_chunks;
+    }
+    return result;
+}
 
-        BOOST_REQUIRE(error == NULL);
+uint32_t getFileIndexForChunk(uint64_t* max_file_index, ds3_str* current_obj_name, checksum_result* results) {
+    uint64_t k;
```
Comment (DenverM80) [inline test/bulk_get.cpp]: I prefer better named variables, others probably won't care

File: test/bulk_get.cpp [hunk 4]
```diff
@@ -7,153 +7,263 @@
...
+        if (g_strcmp0(current_obj_name->value, results[k].original_name) == 0) {
+            file_index = k;
+            break;
+        }
+    }
+    if (file_index == -1) {
+        file_index = *max_file_index;
+        (*max_file_index)++;
+                
+        memcpy(results[file_index].tmp_name+5, current_obj_name->value, 64-5);
```
Comment (DenverM80) [inline test/bulk_get.cpp]: more magic numbers that could be defined as static variables

File: test/bulk_get.cpp [hunk 5]
```diff
@@ -7,153 +7,263 @@
...
+        
+        memcpy(results[file_index].original_name, current_obj_name->value, 64);
+    }
+    results[file_index].num_chunks++;
+    return file_index;
+}
 
-        BOOST_REQUIRE(chunk_response != NULL);
+void checkChunkResponse(ds3_client* client, uint32_t num_files, ds3_get_available_chunks_response* chunk_response, checksum_result* results) {
+    uint64_t i, n;
```
Comment (DenverM80) [inline test/bulk_get.cpp]: object_list_index?
chunk_object_size?

File: test/bulk_get.cpp [hunk 6]
```diff
@@ -241,26 +378,26 @@ BOOST_AUTO_TEST_CASE( partial_get ) {
...
                                         "%601234567890-%3D~%21%40%23%24%25%5E%26%2A%28%29_%2B%5B%5D%7B%7D%7C%3B%3A%2C./%3C%3E%3F"};
-    const char *range_header_results[5]={"some%20normal%20text", "%2Fan%2Fobject%2Fname", "bytes=0-255,300-400,550-800", "orqwerty%2Fqwerty%40qwerty=",
+    const char *range_header_results[5] = {"some%20normal%20text", "%2Fan%2Fobject%2Fname", "bytes=0-255,300-400,550-800", "orqwerty%2Fqwerty%40qwerty=",
                                          "%601234567890-=~%21%40%23%24%25%5E%26%2A%28%29_%2B%5B%5D%7B%7D%7C%3B%3A,.%2F%3C%3E%3F"};
-    const char *general_delimiter_results[5]={"some%20normal%20text", "/an/object/name", "bytes=0-255%2C300-400%2C550-800", "orqwerty/qwerty@qwerty=",
+    const char *general_delimiter_results[5] = {"some%20normal%20text", "/an/object/name", "bytes=0-255%2C300-400%2C550-800", "orqwerty/qwerty@qwerty=",
                                               "%601234567890-=~%21@%23%24%25%5E%26%2A%28%29_%2B%5B%5D%7B%7D%7C%3B%3A%2C./%3C%3E%3F"};
 
-    for(int i=0; i<5; i++) {
+    for (int i = 0; i < 5; i++) {
```
Comment (DenverM80) [inline test/bulk_get.cpp]: apparently these tests are C++, so we can declare loop variables in the for(), which I would go ahead and do above

File: test/test.cpp [hunk 7]
```diff
@@ -73,14 +73,14 @@ void clear_bucket(const ds3_client* client, const char* bucket_name) {
...
-	    ds3_free_request(request);
-	}else{
-	    request = ds3_init_delete_object(bucket_name, bucket_response->objects[i].name->value);
-	    error = ds3_delete_object(client, request);
-	    ds3_free_request(request);
-	}
+            request = ds3_init_delete_folder(bucket_name, bucket_response->objects[i].name->value);
+            error = ds3_delete_folder(client, request);
+            ds3_free_request(request);
+        }else{
```
Comment (DenverM80) [inline test/test.cpp]: } else {

in general, its nice to see you fixing these formatting and style inconsistencies!

File: test/test.cpp [hunk 8]
```diff
@@ -100,90 +100,115 @@ void populate_with_objects(const ds3_client* client, const char* bucket_name) {
...
-      ds3_free_request(request);
+ds3_get_available_chunks_response* ensure_available_chunks(const ds3_client* client, ds3_str* job_id) {
+    ds3_request* request = NULL;
+    ds3_error* error = NULL;
+    bool retry_get;
+    ds3_get_available_chunks_response* chunk_response;
+    do {
+        retry_get = false;
+        request = ds3_init_get_available_chunks(job_id->value);
+
```
Comment (DenverM80) [inline test/test.cpp]: whitespace (blank lines) are kinda arbitrary here

File: src/ds3.c [hunk 9]
```diff
@@ -453,6 +453,23 @@ void ds3_request_set_max_keys(ds3_request* _request, uint32_t max_keys) {
     _set_query_param(_request, "max-keys", max_keys_s);
 }
 
+const char* UNSIGNED_LONG_BASE_10 = "4294967296";
```
Comment (rpmoore) [inline src/ds3.c]: I would make these static so that they are not exported in the symbol table and stay local to the file.

**Pull request merged**